### PR TITLE
Bring old seed technique back

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR disables the seedable random sampler by default. While yes it can make user results *fully reproducible*, it has a questionable degradation in quality which outweighs its benefits. It can be enabled explicitly in the `Accelerator`. 

Fixes https://github.com/huggingface/accelerate/issues/2268
Fixes https://github.com/huggingface/accelerate/issues/2208

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@BenjaminBossan

### CODE REVIEW

**Comment by muellerzr** [comment]
re-done version of https://github.com/huggingface/accelerate/pull/2264, rebase issues 😓

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -217,6 +217,10 @@ class Accelerator:
             If set to `True`, in cases where the total batch size across all processes does not exactly divide the
             dataset, samples at the start of the dataset will be duplicated so the batch can be divided equally among
             all workers.
+        use_seedable_sampler (`bool`, *optional*, defaults to `False`):
+            Whether or not use a fully seedable random sampler ([`~data_loader.SeedableRandomSampler`]). Comes at a
+            cost of potentially different performances due to different shuffling algorithms, but will ensure the
+            training results are fully reproducible.
```
Comment (BenjaminBossan) [inline src/accelerate/accelerator.py:223]: Should it mention that users should manually `set_seed` if they enable this?

File: src/accelerate/test_utils/scripts/test_script.py [hunk 2]
```diff
@@ -360,18 +363,28 @@ def mock_training(length, batch_size, generator):
...
+    train_set, old_model = mock_training(length, batch_size * state.num_processes, generator, use_seedable_sampler)
     assert are_the_same_tensors(old_model.a), "Did not obtain the same model on both processes."
     assert are_the_same_tensors(old_model.b), "Did not obtain the same model on both processes."
 
     accelerator = Accelerator()
-    train_dl = DataLoader(train_set, batch_size=batch_size, shuffle=True, generator=generator)
+    if use_seedable_sampler:
+        # The SeedableRandomSampler is needed during distributed setups
+        # for full reproducability across processes with the `DataLoader`
+        sampler = SeedableRandomSampler(
```
Comment (BenjaminBossan) [inline src/accelerate/test_utils/scripts/test_script.py:380]: I don't quite understand, shouldn't the `SeedableRandomSampler` be set automatically when `use_seedable_sampler=True`?

**Review by BenjaminBossan — APPROVED** [review]
In principle, this LGTM. I added some comments, but no blockers for merging.

Regarding the original issue(s), do we have confirmation that the usage of `SeedableRandomSampler` causes problems for the _validation/test scores_ when the _seed is fixed correctly_?

Even if not, I can see the argument for going back to the previous default for conserving previous behavior, but arguably it should be recommended to use this new setting then.

File: src/accelerate/accelerator.py (hunk 1 above)
Comment (muellerzr) [reply src/accelerate/accelerator.py:223]: Yes that would definitely be good. Will do so

**Comment by vanbasten23** [comment]
Hi @muellerzr , we observed that the test `accelerate/test_utils/scripts/test_script.py` started failing on cloud TPU v3 (but not v4) after this commit ([failure](https://gist.github.com/vanbasten23/4a1ddddf615b4e0d78767ae936e6fb25)). Per https://github.com/huggingface/transformers/pull/28421, it sounds to me the plan is to make Trainer take an accelerate configuration ( Need to set `seedable_sampler=True`), right?
cc: @will-cromar @JackCaoG

**Comment by muellerzr** [comment]
Correct. I’ll be doing that this week/next

**Comment by vanbasten23** [comment]
hi @muellerzr , just want to follow up and check if we can make Trainer take an accelerate configuration so to make our test pass on cloud TPU v3.

**Comment by muellerzr** [comment]
@vanbasten23 almost. See the PR for it here: https://github.com/huggingface/transformers/pull/28664

**Comment by vanbasten23** [comment]
Hi @muellerzr , could you check if https://github.com/huggingface/transformers/pull/28664 is ready to merge? We would like to build on top your PR and bring the TPU tests into a successful state. Thanks!
